elements/dash-tab-panel/dash-tab-panel.js

```diff
@@ -4,174 +4,217 @@
+   *
+   * Children of the `TabList` with `role=tabpanel` attribute are panels.
    */
-  connectedCallback() {
-    // Acquire all tabs and panels inside the element
-    const tabs = Array.from(this.querySelectorAll('[aria-role=tab]'));
-    const panels = Array.from(this.querySelectorAll('[aria-role=tabpanel]'));
-    // If there are no tabs, there is no way to switch between panels. Abort.
-    if (tabs.length === 0) return;
-
-    // For progressive enhancement, the markup should alternate between tabs
-    // and panels. If JavaScript is disabled, all panels are
-    // visible with their respective tab right above them.
-    // If JavaScript is enabled, the element reorders all children into
-    // their two groups. First all the tabs, then all the panels.
-    // Calling `appendChild` on an already inserted element _moves_ the
-    // element to the last child position.
-    tabs.forEach(tab => this.appendChild(tab));
-    panels.forEach(panel => this.appendChild(panel));
-
-    // The element checks if any of the tabs have been marked as selected. If
-    // not, the first tab is now selected.
-    const selectedTab =
-      tabs.find(e => e.getAttribute('aria-selected') === 'true') || tabs[0];
-
-    // Next, we switch to the selected tab. `selectTab` takes care of marking
-    // all other tabs as deselected and hiding all other panels.
-    this._selectTab(selectedTab);
-  }
+  class TabList extends HTMLElement {
+
+    /**
+     * The constructor hooks up the event handlers. Since only events on the
+     * element itself are handled, this can take place in the constructor.
+     */
+    constructor() {
+      super();
+    }
 
-  /**
-   * `reset` marks all tabs as deselected and hides all the panels.
-   */
-  reset() {
-    const tabs = Array.from(this.querySelectorAll('[aria-role=tab]'));
-    const panels = Array.from(this.querySelectorAll('[aria-role=tabpanel]'));
-
-    tabs.forEach(tab => {
-      tab.classList.remove('selected');
-      tab.setAttribute('tabindex', '-1');
-      tab.setAttribute('aria-selected', 'false');
-    });
-
-    panels.forEach(panel => {
-      panel.classList.add('hidden');
-      panel.setAttribute('aria-hidden', 'true');
-    });
-  }
+    /**
+     * `connectedCallback` groups tabs and panels by reordering and makes sure
+     * exactly one tab is active.
+     */
+    connectedCallback() {
+      // The element needs to do some manual input event handling to allow
+      // switching with arrow keys and Home/End.
+      this.addEventListener('keydown', this._onKeyDown);
+      this.addEventListener('click', this._onClick);
+
+      this.setAttribute('role', 'tablist');
+      // Acquire all tabs and panels inside the element
+      const tabs = Array.from(this.querySelectorAll('[role=tab]'));
+      const panels = Array.from(this.querySelectorAll('[role=tabpanel]'));
+      // If there are no tabs, there is no way to switch between panels. Abort.
+      if (!tabs.length) return;
```

Comment:
On second thought, I think your approach of not relying on falsiness is probably better. It's also more explicit which is good for teaching stuff. I vote we go back to !== 0